### ISSUE TITLE
how to use SelectListView optionKeys

### ISSUE THREAD

**Issue by mnsandler** [HomeSeer/Plugin-SDK#136]
given the below Constructor, how/where would i save the optionKey so i can reuse it later to select the default selection?

Public Sub New(id As String, name As String, options As List(Of String), optionKeys As List(Of String), style As ESelectListType = ESelectListType.DropDown, selection As Integer = -1)

I need to rely on my 'optionKey' not the 'selection' property as the list of options might change (eg a list of devices)

**Comment by alexbk66** [comment]
```
/// <summary>
/// Helper for updaing SelectListView mapping
/// </summary>
/// <param name="currentView"></param>
/// <param name="changedView"></param>
void UpdateSelect(SelectListView currentView, SelectListView changedView)
{
    // changedView contains just numers (as strings)
    // get the string from original view
    // i.e. "3"
    string id = currentView.Id;
    string selected = changedView.Options[changedView.Selection];
    if (int.TryParse(selected, out int i))
    {
        selected = currentView.OptionKeys?[i] ?? currentView.Options[i];
        viewsValueMap.Update(id, selected);
    }
}

```

**Comment by mnsandler** [comment]
Alex,
thanks for the snippet. you can expand on the following line of code? is this saving off the optionkey somewhere?

viewsValueMap.Update(id, selected);

**Comment by alexbk66** [comment]
There's a bit of code behind.  I always create a map which maps UI control to some object property.  So **I don't need to write any code** for handling UI selection.  I just add a map entry like this:
```
pb = new PageBuilderAction(this);
controller.GetChromecastList(out List<string> names, out List<string> keys);
pb.AddSelectListView(deviceLabel, names, keys, this, nameof(Selected));
```
And the user selection is handled by the map - it will update "Selected" property when user selects from the list.
The library was complete for HS3, but now for HS4 it's work in progress.  

I will PM you the code in HS forum.

**Comment by alexbk66** [comment]
```
/// <summary>
/// Add SelectListView and map to target.propertyName
/// </summary>
/// <param name="label">Text label to display</param>
/// <param name="options">Array of strings</param>
/// <param name="keys">A list of keys that corresponds to the Options list</param>
/// <param name="target"></param>
/// <param name="propertyName"></param>
/// <param name="propertyType"></param>
/// <param name="dropdown">ESelectListType.DropDown or ESelectListType.RadioList</param>
/// <returns>View Id</returns>
public string AddSelectListView(string label, IEnumerable<string> options, IEnumerable<string> keys, object target, string propertyName, Type propertyType = null, bool dropdown = true)
{
    string id = GetViewId(label);
    if (propertyType == null)
        propertyType = typeof(string);

    // Add id => target.propertyName mapping, returns current value of target.propertyName
    string selected = viewsValueMap.Add<string>(id, target, propertyName, propertyType);

    List<string> _options = options.ToList();

    ESelectListType style = dropdown ? ESelectListType.DropDown : ESelectListType.RadioList;

    SelectListView view;
    if (keys == null)
    {
        int sel = _options.FindIndex((string s) => s == selected);
        view = new SelectListView(id, label, _options, selection: sel, style: style);
    }
    else
    {
        List<string> _keys = keys.ToList();
        int sel = _keys.FindIndex((string s) => s == selected);
        view = new SelectListView(id, label, _options, _keys, selection: sel, style: style);
    }
    pf.WithView(view);
    return view.Id;
}
```
```
/// <summary>
/// Combine different types of ViewValueMap into one
/// to make map free type (currently string and bool)
/// </summary>
public class AllViewsValueMap
{
    ViewValueMap<string> stringValueMap = new ViewValueMap<string>();
    ViewValueMap<bool> boolValueMap = new ViewValueMap<bool>();


    /// <summary>
    /// Add new entry to map
    /// </summary>
    /// <typeparam name="TProperty">Type of Property</typeparam>
    /// <param name="name"></param>
    /// <param name="target"></param>
    /// <param name="propertyName"></param>
    /// <param name="propertyType">Type of the property to convert from TProperty</param>
    /// <returns></returns>
    public TProperty Add<TProperty>(string name, object target, string propertyName, Type propertyType = null)
    {
        Type type = typeof(TProperty);
        if (propertyType == null)
            propertyType = type;

        object val = null;

        if (type == typeof(bool))
        {
            val = boolValueMap.Add(name, target, propertyName, propertyType);
        }
        else if (type == typeof(string))
        {
            val = stringValueMap.Add(name, target, propertyName, propertyType);
        }
        else
        {
            throw new ArgumentException($"Not supported type '{type.Name}'", "TProperty");
        }
        return val!=null? (TProperty)val : default(TProperty);
    }


    /// <summary>
    /// Set target.propertyName to value
    /// </summary>
    /// <typeparam name="TProperty"></typeparam>
    /// <param name="name"></param>
    /// <param name="value"></param>
    public void Update<TProperty>(string name, TProperty value)
    {
        var type = typeof(TProperty);
        object val = (object)value;

        if (type == typeof(bool))
        {
            boolValueMap.Update(name, (bool)val);
        }
        else if (type == typeof(string))
        {
            stringValueMap.Update(name, (string)val);
        }
        else
        {
            throw new ArgumentException($"Not supported type '{type.Name}'", $"{value}");
        }
    }
}
```

**Comment by alexbk66** [comment]
```
/// <summary>
/// Map view to target.propertyName<TProperty>
/// </summary>
/// <typeparam name="TProperty"></typeparam>
[Serializable]
public class ViewValueMap<TProperty> : Dictionary<string, ViewValueMapEntry<TProperty>>
{
    protected ViewValueMap(System.Runtime.Serialization.SerializationInfo serializationInfo, System.Runtime.Serialization.StreamingContext streamingContext)
    {
        throw new NotImplementedException();
    }

    public ViewValueMap()
    {
    }


    /// <summary>
    /// Add new entry to map
    /// </summary>
    /// <param name="name"></param>
    /// <param name="target"></param>
    /// <param name="propertyName"></param>
    /// <param name="propertyType"></param>
    /// <returns></returns>
    public TProperty Add(string name, object target, string propertyName, Type propertyType = null)
    {
        this[name] = new ViewValueMapEntry<TProperty>(target, propertyName, propertyType);
        return this[name].GetVal();
    }


    /// <summary>
    /// Set target.propertyName to value
    /// </summary>
    /// <param name="name"></param>
    /// <param name="value"></param>
    public void Update(string name, TProperty value)
    {
        Debug.Assert(this.ContainsKey(name));
        if(!this.ContainsKey(name))
            throw new KeyNotFoundException($"Unknown View ID: {name}");

        this[name].SetVal(value);
    }
}
```

**Comment by mnsandler** [comment]
Alex,
thanks for reply and code samples. i think i have a slightly different use case than the need to remap old options/keys to new options/keys.

my insteon plugin uses the insteon id (eg AA:AA:AA) to uniquely identify a device. I generate a list of devices by name for the user to select as part of my Action. The issue is that device AA:AA:AA could change its position in the List as new devices are added or deleted. If the Action only saves the List Index of the option selected (e.g. 1), I can't guarantee index 1 will always point to device AA:AA:AA in subsequent lists. However if i could save AA:AA:AA as part of the Action definition, I could use that value to reselect the device as the default option in the List.

are there any user objects in Triggers or Actions that we can save our own data values?

**Comment by alexbk66** [comment]
I haven't looked at triggeres yet, only actions

**Comment by mnsandler** [comment]
I think I figured out the answer based on some experimenting. The method GetSelectedOptionKey() appears to provide the OptionKey from the initial view definition. One can retrieve this value later for use. Thus you don't need to rely on the view.selection property

Triggers seems to work a little different than Actions. Specifically the lifecycle flow.

I used the method AbstractActionType.OnConfigItemUpdate to modify ConfigPage based on the selected Action, or view data.

The Trigger Class handle subtriggers which are not created from a user view, So AbstractTriggerType.OnConfigItemUpdate doesn't get called if you change the subtrigger selection. So I made all my changes to the Configpage in the New Constructor.
